### PR TITLE
task(providerCleanup): Limit StationsProvider logic to handle only the stations by untroducing a new appStore. FWM-34

### PR DESCRIPTION
**Pull request by dimitrantz**
- Use zustand to manage the state across the application
- Split the StationsProvider logic, keep the Provider only for fetching the stations and move any extra logic to the AppStore 
- Create `weatherDataFormatUtils` and move util methods outside the StationModalContent to keep the component lighter during the rendering. I've split the component into 2 smaller ones just for readability but happy to revert the change if we don't like it.

### CODE REVIEW

File: src/hooks/useAppStore.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { create } from "zustand";
```
Comment (Virtuosofriend) [inline src/hooks/useAppStore.ts:1]: I feel that zustand can be used for multiple store instances. You basically can create different hooks and use them to bundle functionality and then share it with different modules. 
If that's the case, maybe we could rename this to be more stations specific rather than AppStore. If not, ignore my comment.

File: src/components/Home/HomepageMap.tsx [hunk 2]
```diff
@@ -32,21 +26,20 @@ export default function HomepageMap() {
     const minZoom = CONFIG.minZoom;
 
     // Create the state for the modal info
-    const [activeStation, setActiveStation] = useState(0);
-    const [markers, setMarkers] = useState<any>([]);
```
Comment (Virtuosofriend) [inline src/components/Home/HomepageMap.tsx:36]: To be honest I don't remember why had added this in the first place. Due to the flashing of the markers on each click?

File: src/components/Home/HomepageMap.tsx [hunk 3]
```diff
@@ -67,33 +60,21 @@ export default function HomepageMap() {
         });
     };
 
-    useMemo(() => {
```
Comment (Virtuosofriend) [inline src/components/Home/HomepageMap.tsx:70]: Same question as above.

File: src/components/Home/StationModalContent.tsx [hunk 4]
```diff
@@ -6,238 +6,172 @@ import BaseWeatherIcon from "@/components/BaseComponents/BaseWeatherIcon";
...
-                            </div>
+    return (
+        <div>
+            {weatherData.map((elem: ExportedWeatherData) => {
+                return (
+                    <div className="p-2 flex text-black flex-col relative" key={elem.station.id}>
+                        {isLoading && loadingBlock}
+                        <div className="flex flex-col">
+                            <StationWeatherSummary {...elem} />
+                            <StationWeatherDetails {...elem} />
```
Comment (Virtuosofriend) [inline src/components/Home/StationModalContent.tsx:170]: I like this so very much. Much much cleaner approach than before and probably we could move the `StationWeatherSummary` and `StationWeatherDetails` in different file components as well in the future.

Comment (dimitrantz) [reply src/components/Home/StationModalContent.tsx:170]: Brilliant, and yep I agree, I'm happy to move them out now if you agree.

File: src/hooks/useAppStore.ts (hunk 1 above)
Comment (dimitrantz) [reply src/hooks/useAppStore.ts:1]: From what I've found so far, it seems that until a couple years ago the suggestion was to use always and only one single store as the single source of truth for the state of the whole app. Zustand is an unopinionated tool, there are no strict rules, and although its implementation is not blocking having many stores, it is still suggested to keep one main store (zustand's official doc about recommented patterns https://github.com/pmndrs/zustand/blob/main/docs/guides/flux-inspired-practice.md ). When reading through documents and articles (trying not to be too biased), having many stores is not necessary a bad design, especially when the app grows and the store becomes huge. It could help logically group the components' state where needed. However, I'm still not convinced that having many stores is a good design for a few reasons. I've listed some of my thoughts below:
- although the stores can be isolated, there could be overrides/collisions causing issues, there is no way to practically safeguard this
- having multiple stores, doesn't mean necessarily different exposed hooks, they will need to be sliced and then combined back to one main store (at least using zustand)
- it seems the only real benefit would be improving the performance, but as mentioned this comes with slicing the main store, which comes with extra work and effort.

To summarise, personally I'd prefer to keep one store because logically the whole concent is around the station, considering also that our application is very light and small as of now. If we still want to explore the other approach I'm happy to give it a go.

File: src/components/Home/HomepageMap.tsx (hunk 3 above)
Comment (dimitrantz) [reply src/components/Home/HomepageMap.tsx:70]: please find the  reply in the comment here https://github.com/Weath-r/fwm-app/pull/31#discussion_r1489794408

File: src/components/Home/HomepageMap.tsx (hunk 2 above)
Comment (dimitrantz) [reply src/components/Home/HomepageMap.tsx:36]: Yep, I believe that's right. Before the changes we had a bit too much logic in the provider. While exploring and reading more about the re-rendeing of the component in general, we found that when the state coming from the provider changes, it leads to the rendering of any of its children components. For us, it was re-rendering the whole component and that was leading to this flashing. We had found that using the state and memoizing the value, the flashing was resolved. However keeping the provider smaller with specific context I noticed we don't really need this, because indeed if there is a change in the stations, then we should correctly re-render.

File: src/hooks/useAppStore.ts (hunk 1 above)
Comment (Virtuosofriend) [reply src/hooks/useAppStore.ts:1]: Wow, thank you for the detailed answer. I do agree with going forward with one store since the app won't be so big!

**Review by Virtuosofriend — APPROVED** [review]
